### PR TITLE
refactor(common): replace CairoMSerialize with typed ABI codec layer

### PR DESCRIPTION
**Pull request by enitrat**
## Summary

This PR replaces the old `CairoMSerialize` slot-based serialization system with a comprehensive typed ABI codec layer that provides better type safety, validation, and maintainability.

### New ABI Codec System
- **Typed ABI layer** (`abi_codec.rs`): Introduces `AbiType`, `CairoMValue`, and `InputValue` enums for type-safe encoding/decoding
- **Comprehensive validation**: Boolean values (0/1), U32 range checks, trailing data detection
- **CLI argument parser**: Supports nested tuples `(1,2,3)` and structs `{1,{2,3}}`
- **Proper error handling**: `AbiCodecError` type instead of `anyhow` in library code

### Compiler Integration
- MIR types now map directly to ABI types
- Program metadata includes typed parameter/return information via `AbiSlot`
- Entrypoints carry full ABI metadata for type-safe execution

### Runner & Prover Improvements
- Typed value handling throughout execution pipeline
- Improved CLI documentation with examples
- Support for both tuple syntaxes: `(1,2,3)` and `[1,2,3]`
- Differential test harness supports nested structures

### Testing Infrastructure
- **Property-based testing**: Comprehensive proptest round-trip tests for all types
- **Organized test structure**: Separate modules for edge cases, parser, integration, and proptest
- **DRY principle**: Removed redundant tests covered by property testing
- **Test coverage**: 27 tests covering validation, parsing, encoding/decoding, and edge cases

## Why These Changes?

The previous `CairoMSerialize` system used untyped "slot math" which was:
- Error-prone and difficult to debug
- Hard to extend for new types
- Lacking proper validation
- Missing type information at runtime

The new ABI layer provides:
- ✅ Type safety at encode/decode boundaries
- ✅ Better error messages for users
- ✅ Foundation for future features (arrays, more complex types)
- ✅ Cleaner separation between compiler and runtime concerns
- ✅ Maintainable and well-tested codebase

## Notes

- Fixed-size arrays are marked as unsupported with TODO comments (tracked in Linear issue CORE-1118)
- All existing tests pass, including new property-based tests
- Backwards compatible JSON format for compiled programs
- Code follows project conventions and is properly formatted

## Testing

```bash
# Run all ABI codec tests
cargo test -p cairo-m-common abi_codec

# Run differential tests
cargo test -p cairo-m-runner --test diff_tests

# Run proptest suite
cargo test -p cairo-m-common proptest
```

All tests pass ✅

### CODE REVIEW

File: crates/common/src/abi_codec.rs [hunk 1]
```diff
@@ -0,0 +1,923 @@
...
+                        fields,
+                    }
+                ),
+            ]
+            .boxed()
+        }
+
+        proptest! {
+            #[test]
+            fn felt_roundtrip(n in 0i64..=i64::MAX) {
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:811]: Why only positive numbers for i64?

File: crates/common/src/abi_codec.rs [hunk 2]
```diff
@@ -0,0 +1,923 @@
...
+                Some(b't') | Some(b'f') => self.parse_bool(),
+                Some(b'-') | Some(b'0'..=b'9') => self.parse_number(),
+                _ => Err(AbiCodecError::ParseError(format!(
+                    "Unexpected character '{}' at position {}. Expected number, boolean, tuple '(' or '[', or struct '{{'",
+                    self.peek().map(|b| b as char).unwrap_or('\0'),
+                    self.i
+                ))),
+            }
+        }
+        fn parse_array(&mut self) -> Result<InputValue, AbiCodecError> {
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:363]:
```suggestion

        fn parse_array(&mut self) -> Result<InputValue, AbiCodecError> {
```

File: crates/common/src/abi_codec.rs [hunk 3]
```diff
@@ -0,0 +1,923 @@
...
+        i: usize,
+    }
+    impl<'a> Parser<'a> {
+        const fn new(s: &'a str) -> Self {
+            Self {
+                src: s.as_bytes(),
+                i: 0,
+            }
+        }
+        const fn eof(&self) -> bool {
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:326]:
```suggestion

        const fn eof(&self) -> bool {
```

File: crates/common/src/abi_codec.rs [hunk 4]
```diff
@@ -0,0 +1,923 @@
...
+            self.i >= self.src.len()
+        }
+        fn peek(&self) -> Option<u8> {
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:329]:
```suggestion

        fn peek(&self) -> Option<u8> {
```

File: crates/common/src/abi_codec.rs [hunk 5]
```diff
@@ -0,0 +1,923 @@
...
+            self.src.get(self.i).copied()
+        }
+        fn next(&mut self) -> Option<u8> {
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:332]:
```suggestion

        fn next(&mut self) -> Option<u8> {
```

File: crates/common/src/abi_codec.rs [hunk 6]
```diff
@@ -0,0 +1,923 @@
...
+            Ok((CairoMValue::Array(out), cur))
+        }
+        AbiType::Unit => Ok((CairoMValue::Unit, start)),
+    }
+}
+
+/// Parse a CLI argument string into an InputValue (supports nesting)
+///
+/// Supported grammar (positional structs):
+///   Value := Number | Bool | Array | Struct
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:311]: tuple should be added?

File: crates/common/src/abi_codec.rs [hunk 7]
```diff
@@ -0,0 +1,923 @@
...
+///   Array := '[' (Value (',' Value)*)? ']'
+///   Struct := '{' (Value (',' Value)*)? '}'
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:313]:
```suggestion
///   Value := Number | Bool | Array | Tuple | Struct
///   Array := '[' (Value (',' Value)*)? ']'
///   Tuple := '(' (Value (',' Value)*)? ')'
///   Struct := '{' (Value (',' Value)*)? '}'
```

File: crates/common/src/abi_codec.rs [hunk 8]
```diff
@@ -0,0 +1,923 @@
...
+                    }
+                    _ => {
+                        return Err(AbiCodecError::ParseError(format!(
+                            "Expected ',' or ']' at position {} in tuple",
+                            self.i
+                        )))
+                    }
+                }
+            }
+            Ok(InputValue::List(items))
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:391]: could be factorized between parse_array, parse_tuple, parse_struct

File: crates/common/src/abi_codec.rs [hunk 9]
```diff
@@ -0,0 +1,923 @@
...
+                        self.next();
+                        self.skip_ws();
+                    }
+                    Some(b']') => {
+                        self.next();
+                        break;
+                    }
+                    _ => {
+                        return Err(AbiCodecError::ParseError(format!(
+                            "Expected ',' or ']' at position {} in tuple",
```
Comment (obatirou) [inline crates/common/src/abi_codec.rs:385]:
```suggestion
                            "Expected ',' or ']' at position {} in array",
```
